smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,251 @@
+                openBlock(
+                    "guard let serviceParams = params as? \$L else {",
+                    "}",
+                    serviceParamsName
+                ) {
+                    write("throw ClientError.authError(\"Service specific auth scheme parameters type must be passed to auth scheme resolver.\")")
+                }
+
+                renderSwitchBlock(serviceIndex, ctx, this)
+            }
+        }
+    }
+
+    private fun renderSwitchBlock(
+        serviceIndex: ServiceIndex,
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter
+    ) {
+        writer.apply {
+            // Switch block for iterating over operation name cases
+            openBlock("switch serviceParams.operation {", "}") {
+                // Handle each operation name case
+                val operations = ctx.service.operations
+                operations.forEach {
+                    val opShape = ctx.model.getShape(it).get() as OperationShape
+                    if (
+                        opShape.hasTrait(AuthTrait::class.java) ||
+                        opShape.hasTrait(OptionalAuthTrait::class.java) ||
+                        opShape.hasTrait(UnsignedPayloadTrait::class.java)
+                    ) {
+                        val opName = it.name.toLowerCamelCase()
+                        val sdkId = getSdkId(ctx)
+                        val validSchemesForOp = serviceIndex.getEffectiveAuthSchemes(
+                            ctx.service,
+                            it,
+                            ServiceIndex.AuthSchemeMode.NO_AUTH_AWARE
+                        )
+                        renderOperationSwitchCase(sdkId, opShape, opName, validSchemesForOp, writer)
+                    }
+                }
+                // Handle default case, where operations default to auth schemes defined on service shape
+                val validSchemesForService = serviceIndex.getEffectiveAuthSchemes(ctx.service, ServiceIndex.AuthSchemeMode.NO_AUTH_AWARE)
+                renderDefaultSwitchCase(getSdkId(ctx), validSchemesForService, writer)
+            }
+
+            // Return result
+            write("return validAuthOptions")
+        }
+    }
+
+    private fun renderOperationSwitchCase(sdkId: String, opShape: OperationShape, opName: String, schemes: Map<ShapeId, Trait>, writer: SwiftWriter) {
+        writer.apply {
+            write("case \"$opName\":")
+            indent()
+            schemes.forEach {
+                if (it.key == SigV4Trait.ID) {
+                    write("var sigV4Option = AuthOption(schemeID: \"${it.key}\")")
+                    write("sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: ${(it.value as SigV4Trait).name})")
+                    openBlock("guard let region = serviceParams.region else {", "}") {
+                        write("throw ClientError.authError(\"Missing region in auth scheme parameters for SigV4 auth scheme.\")")
```

Comment:
Should consider making various error messages constants rather than in-line during the codegen string so that it is more readable + potentially reusable.

Reply:
Changed as suggested.